bitsandbytes/backends/__init__.py

```diff
@@ -0,0 +1,8 @@
+from typing import Dict
+
+from bitsandbytes.backends.base import Backend
+
+backends: Dict[str, Backend] = {}
```

Comment:
I wonder if backends should be private?

Reply:
Why should it be?

Reply:
Does it intend to avoid the usage of an `import ` of the `backends` from the bitsandbytes package (and we may make it `_backends`) ?

Reply:
How does a user currently select a backend? Currently, only CUDA is supported, but should there be a function like `backend.set_backend("cuda")`?

I guess we would do this through the "device_setup" process. The question here is if we can automatically detect the device the user is running in all cases? I think the only exception is probably if a user has both an accelerated device and a CPU. I think having, for example, Apple silicon and a regular GPU will not really happen. Are there any other scenarios that we are missing here and we need to think about?

I think for now, it looks fine, but I want to make sure we are not missing anything. In terms of usability, the best designs often come from early thought rather than later corrections. So it makes sense we think a bit about this.

Reply:
@TimDettmers To your point, I wonder if IPEX can be combined with CUDA/ROCm in such a way where as you mention, it's not clear what the user will want.

E.g. a situation where both `torch.xpu.is_available() == true` and `torch.cuda.is_available() == true`.

It's also my understanding that Intel GPU support may be upstreamed: pytorch/pytorch#114842

Reply:
@albanD from PyTorch was also raising the possibility of Apple Silicon + dedicated GPU. Seems it's something that is a relevant edge case for them. Not sure if this should be relevant given that it's so niche, but if a bit of thought can help with a more generally applicable solution, why not? I agree with Tim that once this interface is in place, it will be very hard to change.

I mentioned because we already have that function of registering the backends, which currently only adds to the dict but which is a natural place to add hooks. That would only be possible of course if we make it clear to the user that they shouldn't interact directly with the backends dict.

Reply:
Since the backends are also registered with the device tag ("cuda", "xpu" etc.), we should be able to decide what backend implementation to invoke according to the devices of the tensor args?

Reply:
@jgong5 That's what's happening now in this PR 😁 

[`return backends[A.device.type].dequantize_4bit(...)`](https://github.com/TimDettmers/bitsandbytes/pull/898/files#diff-f8dbfa8ee0162e506a70aeb3be4f1c29ff19e71d331f184f2b01267de2232dc5R873-R874)

Reply:
Hey all, so Tim thinks that the backend should only be initialized once and therefore implemented as Singleton: It
shouldn't be initialized multiple times through a dictionary.

According to him, there's no use-case to exchange the backend at runtime. The only potential use-case might be that of having both a CPU and GPU backend at the same time, but from what Tim says, this is sth that we currently don't need yet and shouldn't worry about.

Just forwarding his statement for the sake of furthering the discussion.

Reply:
> From an engineering standpoint, I disagree with implementing it as a singleton (a class you can ever only initialize once). Doing that is more complex, a little non-Pythonic, and the current implementation has the same end result: there's a backend object that's only created once, and it's plugged into place in the backends dict.

I feel the same - no obvious benefit of constraining us with a single device. May I know what's the concern with dispatching device backend from the backend dict with the device on the tensor args? Dispatching according to the tensor's device type is something PyTorch ATen is also doing.

Reply:
Would it not make sense to try to stay with the device of the source/destination tensors rather than select and initialize the device once as a singleton?

If you have multiple GPUs for example and want to share the compute with them, wouldn't you want to do .to(some_device) then call BnB?